feat(kuma-cp) - set host network var as true by default

SallyBlichWalkMe opened the pull request:
### Summary

For alternate overlay networking(other than the basic CNI provided by EKS for example, like cilium), 
This is mandatory value that has to be set, otherwise it will have communication errors with the control plane.
If you are using the default CNI component, it's running on the default network.

### Full changelog

* Implement hostNetwork: true in cp-deployment.yaml file.

### Documentation

- [ ] Link to the website [documentation PR](https://github.com/kumahq/kuma-website/pull/XXX)

### Testing

- [ ] Unit tests
- [ ] E2E tests
- [ ] Manual testing on Universal
- [ X ] Manual testing on Kubernetes

### Backwards compatibility

- [ ] Update [`UPGRADE.md`](/UPGRADE.md) with any steps users will need to take when upgrading.
- [ ] Add `backport-to-stable` label if the code follows our [backporting policy](/CONTRIBUTING.md#backporting)

jakubdyszkiewicz commented:
Hey, thanks for your contribution!

I'd like to understand why this value has to be set to true in order for Cilium (and others) to work. Is there any doc on that?

The doc on Kubernetes side is also not very descriptive about potential downsides.

> Host networking requested for this pod. Use the host's network namespace. If this option is set, the ports that will be used must be specified. Default to false.

I'm a bit concerned about the consequences of this change. Could we make this an opt-in? Meaning that let's create `controlPlane.hostNetworking` in `Values.yaml` with default value false and you can enable true in your deployment.

SallyBlichWalkMe commented:
Hi, @jakubdyszkiewicz  thanks for the answer.
The issue that I had, was with the dataplanes, they couldn't be able to connect to the control plane, until I set inside the control plane's deployment file the attribute of hostNetwork: true, because the default mash has not been created.
Thats the same issue that I was having: https://github.com/kumahq/kuma/issues/2167
Regrading the proposal that you suggested - while I be able to pass it to kuma's helm chart as var if so?

jakubdyszkiewicz commented:
> The issue that I had, was with the dataplanes, they couldn't be able to connect to the control plane, until I set inside the control plane's deployment file the attribute of hostNetwork: true, because the default mash has not been created.

I think the question is why the default mesh cannot be created? I see in the issue #2167 that Kubernetes cannot call Kuma webhook. It would be great to know why.

> Regrading the proposal that you suggested - while I be able to pass it to kuma's helm chart as var if so?

Do you install Kuma with HELM or with kumactl?

If you add a value to HELM called `controlPlane.hostNetworking`, you just do
```
helm install --create-namespace --namespace --set controlPlane.hostNetworking=true kuma-system kuma kuma/kuma
```

In addition to that we can add a kumactl flag called `--control-plane-host-networking` so if you install it via kumactl you can do
```
kumactl install control-plane --control-plane-host-networking | kubectl apply -f -
```

jakubdyszkiewicz commented:
Ok cool, let's add this new config to `Values.yaml` then

SallyBlichWalkMe commented:
> Ok cool, let's add this new config to `Values.yaml` then

great, I added the attribute and removed it from cp-deployment.yaml

jakubdyszkiewicz commented:
@SallyBlichWalkMe 
we need to now use this variable so in `cp-deployment.yaml` there should be
```
  hostNetwork: {{ .Values.controlPlane.hostNetwork }}
```
please remember to run 
```
make test PKG_LIST=./app/kumactl/cmd/install/... UPDATE_GOLDEN_FILES=true
```
because I think this will break tests.

Also, we need DCO :) https://github.com/kumahq/kuma/pull/3992/checks?check_run_id=5569918916

SallyBlichWalkMe commented:
> 
> ```
> make test PKG_LIST=./app/kumactl/cmd/install/... UPDATE_GOLDEN_FILES=true
> ```
> 
> because I think this will break tests.

Thanks for the help,
I added the attribute where you said, and added DCO.
Regarding the tests, I tried to run it from kuma's directory and it failed:

```
➜  kuma git:(feat/set_host_network_varmakemake test PKG_LIST=./app/kumactl/cmd/install/... UPDATE_GOLDEN_FILES=true
/bin/sh: go: command not found
find: /pkg/mod/github.com/envoyproxy: No such file or directory
find: /pkg/mod/github.com/cncf: No such file or directory
find: /pkg/mod/github.com/googleapis: No such file or directory
find: /pkg/mod/github.com/kumahq: No such file or directory
make: go: Command not found
make: go: Command not found
mkdir -p "/Users/XXX/Documents/GitHub/kuma/build/coverage"
TMPDIR=/tmp UPDATE_GOLDEN_FILES=true go test  -ldflags="-s -w  -X github.com/kumahq/kuma/pkg/version.version=dev-8f01a5ed  -X github.com/kumahq/kuma/pkg/version.gitTag=1.5.0-rc1-86-g8f01a5ed  -X github.com/kumahq/kuma/pkg/version.gitCommit=8f01a5edaf5c07d46f0a242b6f770cc8db95d6ed  -X github.com/kumahq/kuma/pkg/version.buildDate=2022-03-17T12:13:14Z "  -race -covermode=atomic -coverpkg=./... -coverprofile="/Users/XXX/Documents/GitHub/kuma/build/coverage/coverage.out" $(go list ./app/kumactl/cmd/install/... | grep -E -v "test/e2e" | grep -E -v "pkg/transparentproxy/istio/tools")
/bin/sh: go: command not found
/bin/sh: go: command not found
make: *** [test] Error 127
```


What should I do?

jakubdyszkiewicz commented:
Follow the https://github.com/kumahq/kuma/blob/master/DEVELOPER.md guide
make sure you have dependencies installed and then execute `make dev/tools`

SallyBlichWalkMe commented:
> Follow the https://github.com/kumahq/kuma/blob/master/DEVELOPER.md guide make sure you have dependencies installed and then execute `make dev/tools`

Thanks, I installed what what was missing, executed `make dev/tools` and after that 'make test PKG_LIST=./app/kumactl/cmd/install/... UPDATE_GOLDEN_FILES=true'

That's what I got now:
➜  ```
kuma git:(feat/set_host_network_var) ✗ make chectest PKG_LIST=./app/kumactl/cmd/install/... UPDATE_GOLDEN_FILES=true
TMPDIR=/tmp UPDATE_GOLDEN_FILES=true go test  -ldflags="-s -w  -X github.com/kumahq/kuma/pkg/version.version=dev-8f01a5ed  -X github.com/kumahq/kuma/pkg/version.gitTag=1.5.0-rc1-86-g8f01a5ed  -X github.com/kumahq/kuma/pkg/version.gitCommit=8f01a5edaf5c07d46f0a242b6f770cc8db95d6ed  -X github.com/kumahq/kuma/pkg/version.buildDate=2022-03-17T13:17:26Z "  -race -covermode=atomic -coverpkg=./... -coverprofile="/Users/XXX/Documents/GitHub/kuma/build/coverage/coverage.out" $(go list ./app/kumactl/cmd/install/... | grep -E -v "test/e2e" | grep -E -v "pkg/transparentproxy/istio/tools")
ok      github.com/kumahq/kuma/app/kumactl/cmd/install  17.039s coverage: 13.2% of statements in ./...
?       github.com/kumahq/kuma/app/kumactl/cmd/install/context  [no test files]
/Library/Developer/CommandLineTools/usr/bin/make coverage
GOFLAGS='' go tool cover -html="/Users/XXX/Documents/GitHub/kuma/build/coverage/coverage.out" -o "/Users/XXX/Documents/GitHub/kuma/build/coverage/coverage.html"
```

And I didn't see that it triggered the test of kuma-commit again :(

SallyBlichWalkMe commented:
@jakubdyszkiewicz Is there anything else I should run?

jakubdyszkiewicz commented:
Sorry for not getting back to you!

Run `make check` this will regenerate our HELM docs based on your changes

SallyBlichWalkMe commented:
> Sorry for not getting back to you!
> 
> Run `make check` this will regenerate our HELM docs based on your changes

ok I ran that command too and I get this now:

```
for c in ./deployments/charts/*; do \
                if [ -d $c ]; then \
                        helm lint $c; \
                fi \
        done
==> Linting ./deployments/charts/kuma

1 chart(s) linted, 0 chart(s) failed
/Users/xxxx/bin/golangci-lint run --timeout=10m -v
panic: load embedded ruleguard rules: rules/rules.go:13: can't load fmt

goroutine 1 [running]:
github.com/go-critic/go-critic/checkers.init.9()
        github.com/go-critic/go-critic@v0.6.1/checkers/checkers.go:58 +0x4b4
make: *** [golangci-lint] Error 2

```

I also got a file that that needs be added in this directory: docs/cmd/kuma-cp/kuma-cp.md

lahabana commented:
Maybe you need to update the tools with: `make dev/tools` before running `make check` ?

SallyBlichWalkMe commented:
> Maybe you need to update the tools with: `make dev/tools` before running `make check` ?

Yes, Thats what I did, by this order:
1. Executed make dev/tools 
2. Executed 'make test PKG_LIST=./app/kumactl/cmd/install/... UPDATE_GOLDEN_FILES=true'
3. Executed make check

lahabana commented:
@mergifyio update

lahabana commented on deployments/charts/kuma/values.yaml:
```diff
@@ -161,6 +161,11 @@ controlPlane:
       additionalRules: ""
 
+  # -- Specifies if the deployment should be started in hostNetwork mode.
+  hostNetworking: false
```
```suggestion
  hostNetwork: false
```

lahabana reviewed (commented):
I've checked out your change locally. The diffs you are seeing in your make check is because you branch is not up to date. I've used mergifyio to update your branch so if you pull it should work fine.
I also found a minor bug while validating your change. If you are still stuck and you are ok with me pushing the update to the golden files please let me know.

Thanks!

lahabana commented:
Hey @SallyBlichWalkMe planning on doing this still?

SallyBlichWalkMe commented:
@lahabana
Yes, I pulled from master to get the latest commits, I ran again those commands again:
Executed make dev/tools
Executed 'make test PKG_LIST=./app/kumactl/cmd/install/... UPDATE_GOLDEN_FILES=true'
Executed make check
And I accepted your change 
Still the kuma-commit is failing

jakubdyszkiewicz commented:
Please revert changes in tools/releases/dockerfiles/Dockerfile.kumactl

also try
```
rm ~/bin/helm-docs
make dev/install/helm-docs
make check
```

SallyBlichWalkMe commented:
> Please revert changes in tools/releases/dockerfiles/Dockerfile.kumactl
> 
> also try
> 
> ```
> rm ~/bin/helm-docs
> make dev/install/helm-docs
> make check
> ```

Thanks for the reply, I reverted the change, 
I tried those steps and still the kuma-commit failed over the 'check' step again, and also locally I got this again:

```
for c in ./deployments/charts/*; do \
                if [ -d $c ]; then \
                        helm lint $c; \
                fi \
        done
==> Linting ./deployments/charts/kuma

1 chart(s) linted, 0 chart(s) failed
/Users/xxx/bin/golangci-lint run --timeout=10m -v
panic: load embedded ruleguard rules: rules/rules.go:13: can't load fmt

goroutine 1 [running]:
github.com/go-critic/go-critic/checkers.init.9()
        github.com/go-critic/go-critic@v0.6.1/checkers/checkers.go:58 +0x4b4
make: *** [golangci-lint] Error 2

```

jakubdyszkiewicz commented:
It can be either old version of Go (we now use 1.18) or old version of golangci-lint (you can upgrade with `make dev/install/golangci-lint`)
